cpp/src/community/leiden.cuh

```diff
@@ -0,0 +1,163 @@
+    // during each iteration of the loop
+    bool up_down = true;
+
+    while (new_Q > (cur_Q + 0.0001)) {
+      cur_Q = new_Q;
+
+      this->compute_delta_modularity(
+        total_edge_weight, resolution, graph, cluster_hash_v, old_cluster_sum_v, delta_Q_v);
+
+      // Filter out positive delta_Q values for nodes not in the same constraint group
+      thrust::for_each(
+        rmm::exec_policy(this->stream_)->on(this->stream_),
+        thrust::make_counting_iterator(0),
+        thrust::make_counting_iterator(graph.number_of_edges),
+        [d_src_indices, d_dst_indices, d_constraint, d_delta_Q] __device__(vertex_t i) {
+          vertex_t start_cluster = d_constraint[d_src_indices[i]];
+          vertex_t end_cluster   = d_constraint[d_dst_indices[i]];
+          if (start_cluster != end_cluster) d_delta_Q[i] = weight_t{0.0};
+        });
+
+      this->assign_nodes(graph, cluster_hash_v, next_cluster_v, delta_Q_v, up_down);
+
+      up_down = !up_down;
+
+      new_Q = this->modularity(total_edge_weight, resolution, graph, next_cluster_v.data().get());
+
+      if (new_Q > cur_Q) {
+        thrust::copy(rmm::exec_policy(this->stream_)->on(this->stream_),
+                     next_cluster_v.begin(),
+                     next_cluster_v.end(),
+                     this->cluster_v_.begin());
+      }
+    }
+
+    this->timer_stop(this->stream_);
+    return cur_Q;
+  }
+
+  std::pair<int, weight_t> compute(vertex_t *d_cluster_vec, int max_level, weight_t resolution)
+  {
+    int num_level{0};
+
+    weight_t total_edge_weight = thrust::reduce(rmm::exec_policy(this->stream_)->on(this->stream_),
+                                                this->weights_v_.begin(),
+                                                this->weights_v_.end());
+
+    weight_t best_modularity = weight_t{-1};
+
+    //
+    //  Initialize every cluster to reference each vertex to itself
+    //
+    thrust::sequence(rmm::exec_policy(this->stream_)->on(this->stream_),
+                     this->cluster_v_.begin(),
+                     this->cluster_v_.end());
+    thrust::copy(rmm::exec_policy(this->stream_)->on(this->stream_),
+                 this->cluster_v_.begin(),
+                 this->cluster_v_.end(),
+                 d_cluster_vec);
+
+    //
+    //  Our copy of the graph.  Each iteration of the outer loop will
+    //  shrink this copy of the graph.
+    //
+    GraphCSRView<vertex_t, edge_t, weight_t> current_graph(this->offsets_v_.data().get(),
+                                                           this->indices_v_.data().get(),
+                                                           this->weights_v_.data().get(),
+                                                           this->number_of_vertices_,
+                                                           this->number_of_edges_);
+
+    current_graph.get_source_indices(this->src_indices_v_.data().get());
+
+    while (num_level < max_level) {
```

Comment:
Please consider using a `for_each()` or `transform()` (or, `transform_if()`, `std::any_of()`, `std::all_of()`, `std::none_of()`) , due to the possibility of early break) on the sequence `{0,...,max_level-1}` rather than a while loop. 

This would lead to: 

1. better reading and maintenance;
2. flexibility on experimenting or using different execution policies;
3. better preparation for the future: C++17 is around the corner and STL algorithms can be (CPU) parallelized simply by passing different execution policies (`seq`, `par`, `par_unseq`);

Reply:
I'm not sure how I would restructure this loop efficiently in that manner, given the early break.

This loop must be executed sequentially.  The shrink_graph function (the last step in the loop) creates a new graph for the next iteration.

Reply:
You could use `std::find_if` (or `thrust::find_if()` with `counting_iterator` on the `__host__`). `foo()` and `bar()` are equivalent in the example below:

```
template<typename fun_t,
         typename cond_t>
void foo(size_t max_iter, fun_t callable, cond_t condition)
{
    size_t num{0};
    while( num < max_iter) {
        callable();

        if( condition(num) ) break;
        ++num;
    }
    std::cout << "reached..." << num << '\n';
}

template<typename fun_t,
         typename cond_t>
void bar(size_t max_iter, fun_t callable, cond_t condition)
{
    std::vector<size_t> counter(max_iter);
    std::iota(std::begin(counter), std::end(counter), 0l);

    auto pos = std::find_if(std::begin(counter), std::end(counter),
                            [callable, condition](auto num) {
                                callable();
                                return condition(num);
                            });
    size_t num = std::distance(counter.begin(), pos);
    std::cout << "reached..." << num << '\n';
}
```
However, I admit this might be artificial and less readable. Your call here.

Reply:
OK.  I see now.

I don't see a clarity advantage in the `std::find_if` approach for this loop (and the others structured like this in Louvain).  I could see if we might eventually want to use one of the non-sequential execution policies.

I did endeavor to make the entire contents of these while loops fit on a single screen in my editor, which I have found helps in keeping this type of construct clear.

Reply:
No worries. Feel free to resolve.